lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -74,6 +75,118 @@ static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
     return failure();
   }
 
+  // Store a mapping of defname strings to the first external module
+  // that defines it.
+  llvm::StringMap<FExtModuleOp> defnameMap;
+
+  // Verify external modules.
+  for (auto &op : *circuit.getBody()) {
+    if (auto extModule = dyn_cast<FExtModuleOp>(op)) {
+      if (auto defname = extModule.defnameAttr()) {
+        // Check that this extmodule's defname does not conflict with
+        // the symbol name of any module.
+        auto collidingModule = circuit.lookupSymbol(defname.getValue());
+        if (isa_and_nonnull<FModuleOp>(collidingModule)) {
+          auto diag =
+              op.emitOpError()
+              << "attribute 'defname' with value " << defname
+              << " conflicts with the name of another module in the circuit";
+          diag.attachNote(collidingModule->getLoc())
+              << "previous module declared here";
+          return failure();
+        }
+
+        // Find an optional extmodule with a defname collision. Update
+        // the defnameMap if this is the first extmodule with that
+        // defname or if the current extmodule takes no parameters and
+        // the collision does. The latter condition improves later
+        // extmodule verification as checking against a parameterless
+        // module is stricter.
+        auto collidingExtModuleOpt =
+            Optional<FExtModuleOp>(llvm::NoneType::None);
+        {
+          auto value = defnameMap.find(defname.getValue());
+          if (value != defnameMap.end()) {
+            collidingExtModuleOpt = Optional<FExtModuleOp>(value->getValue());
+            if (value->getValue().parameters() && !extModule.parameters())
+              defnameMap.insert_or_assign(defname.getValue(), extModule);
+          } else {
+            defnameMap.insert_or_assign(defname.getValue(), extModule);
+          }
+        }
+
+        // Go to next extmodule if no extmodule with the same defname
+        // was found.
+        if (!collidingExtModuleOpt)
+          continue;
+
+        auto collidingExtModule = collidingExtModuleOpt.getValue();
+
+        // Check that the number of ports is exactly the same.
+        auto ports = new SmallVector<std::pair<StringAttr, FIRRTLType>, 8>;
+        auto collidingPorts =
+            new SmallVector<std::pair<StringAttr, FIRRTLType>, 8>;
```

Comment:
You generally don't want to heap allocate SmallVectors, and this is leaking them because they don't get freed.  Instead, declare them like this:

`SmallVector<std::pair<StringAttr, FIRRTLType>, 8> ports;`

and then drop the stars down below

Reply:
😬 Thanks for catching this. I've been using managed languages for too long...

Do you have a good suggestion on how to avoid this and/or check for it? When I used to write C/C++ in the past, I used to just run everything through `valgrind` as I didn't trust myself to get manual memory management correct. Do you suggest that, have an alternative, or suggest to just not screw this up?

Reply:
Updated in the fixup commit.

Reply:
I think code review is fine.  No worries about not magically knowing everything, we are all learning different things here :-)